preshed/bloom.pyx

```diff
@@ -54,40 +58,125 @@ cdef class BloomFilter:
+        # unlikely but possible with old data
+        bloom_from_bytes_legacy(mem, bloom, data)
+        return
+    pad, ver, hcount, length, seed = struct.unpack("<QQQQQ", data[0:40])
+    if pad !=0:
+        bloom_from_bytes_legacy(mem, bloom, data)
+        return
+    assert ver == 1, "Unknown serialization version"
+
+    bloom.hcount = hcount
+    bloom.length = length
+    # To avoid overflow, just take the low bits. In valid data nothing will be
+    # lost.
+    cdef uint32_t safe_seed = int.from_bytes(seed.to_bytes(8, 'big')[-4:], 'big')
+    bloom.seed = safe_seed
+
+    buflen = length // KEY_BITS
+    if length % KEY_BITS > 0:
+        buflen += 1
+    contents = struct.unpack(f"<{buflen}Q", data[40:])
+    assert buflen > 0, "Tried to allocate an empty buffer"
+    bloom.bitfield = <key_t*>mem.alloc(buflen, sizeof(key_t))
+    for i in range(buflen):
+        bloom.bitfield[i] = contents[i]
+
+
+cdef void bloom_from_bytes_legacy(Pool mem, BloomStruct* bloom, bytes data):
+    # Older versions of this library used the array module with type L for
+    # serialization. Types in array guarantee a minimum size, not an actual
+    # size, and it turns out L is 8 bytes on Linux and most platforms, but 4 on
+    # Windows. 
+
+    # As a separate issue, due to bits/bytes confusion, each container in the
+    # serialized data has only one byte actually used.
+
+    # The code in this function reads in data in the old format and converts it
+    # to the current format losslessly. It also packs the significant bytes into
+    # contiguous memory.
+
+    # on non-Windows platforms
+    unit = "Q"
+    unit_size = 8 # size of container in bytes
+    offset = unit_size * 3
+    hcount, length, seed = struct.unpack("<QQQ", data[0:offset])
+
+    decode_len = length // unit_size # number of units to unpack
+
+    if length != len(data) - 24:
+        # This can happen if the data was serialized on Windows, where the units
+        # were 32bit rather than 64bit.
+        unit = "L"
+        unit_size = 4
+        offset = unit_size * 3
+        hcount, length, seed = struct.unpack("<LLL", data[0:offset])
+
+        # The length was the number bytes in memory. But because of the
+        # platform size issue, the actual serialized bytes is half that.
+        assert length // 2 == len(data) - offset, "Length is invalid"
+
+        decode_len = length // (2 * unit_size)
+
+    bloom.hcount = hcount
+    bloom.length = length
+    cdef uint32_t safe_seed = int.from_bytes(seed.to_bytes(8, 'big')[-4:], 'big')
+    bloom.seed = safe_seed
+
+    # This is tricky - to remove empty space we're going to map bytes into
+    # containers. On Windows or Linux, length is both the number of significant
+    # bits in the bitfield and the number of bytes when the bitfield was in
+    # memory in the old format. In our output, length will be the bitfield
+    # length in bits.
+    buflen = length // KEY_BITS
+    if length % KEY_BITS > 0:
```

Comment:
```
buflen = decode_len if length % KEY_BITS == 0 else decode_len + 1
```

Expresses the relation more clearly. (Though I am not sure why the decode length wouldn't be increased by one as well, since we might need the extra integer there as well?)

Reply:
As mentioned, this is not the same as `decode_len`.

`decode_len` is a number of containers to be used in the call to `unpack`. Because the data was originally serialized as a number of containers, the calculation for `decode_len` should never have a remainder. (The Windows-specific assert checks this; I could add a more general assert too.)

In contrast, the `buflen` is a buffer we're creating now, and the size is based on the number of significant bits in the incoming data. That may not be a multiple of our container size (`KEY_BITS`), which is why we have to check the modulus here and maybe increase the size.